### ISSUE TITLE
Updating RDDs incrementally

### ISSUE THREAD

**Issue by ashrowty** [ooyala/spark-jobserver#79]
Hi all,

Trying to explore use of Spark as a query engine for an OLAP style dashboard. Planning on the persistent store being Cassandra. How do I update the RDDs as the data is constantly refreshing on disk? The UI dashboard needs to show data upto the minute. Is this an appropriate use case for Spark if the underlying data is changing?

Thanks,
Ashish

**Comment by smagadi** [comment]
try using Elastic search

**Comment by epugh** [comment]
@smagadi I agree that Elasticsearch is really good for streaming aggregations, that has been very much their focus.   

Having said that, I find that at times you need a true computational engine, like Spark, to roll up your data.   For example, let's say you are doing an internet of things type application, and you are picking up millions of events, each of them very small increments of time.   Yes, you could put all that into ES, but you are really going to tax the inverted index.   Instead, dump the events into something that is really mean for lots of writes [Cassandra, or other write oriented system], and then run your streaming Spark over them to aggregate to a point where a search based interface easily parses over them.   Plus, if you want to do more sophisticated analytics, like Complex Event Processing, then you need Spark.   

If you didn't want something like Cassandra at all, you could use Spark to recalculate data (http://opensourceconnections.com/blog/2015/09/24/using-dse-to-run-solr-rdd/ or http://chapeau.freevariable.com/2015/04/elasticsearch-and-spark-1-dot-3.html) over a search index as well, but it stills means you are introducing Spark and need a way to run it.

**Comment by ashrowty** [comment]
I ended up implementing a Spark batch job for now that does the heavy
lifting of rollups and segment calculations and then using Solr to do the
basic slicing and dicing on the summarized data. Going into production in
the next couple of weeks.

The current job takes a long time to complete since I am pre-generating the
output space for all possible permutations and combinations. When I have
some time I want to try the JobServer to do a hybrid approach ... where
some of the calcs are predone but others are real time. That way the system
only needs to compute results for specific user requests.

Spark streaming is interesting but probably won't work since the
calculations require a full 1 year worth of data to compute segments and
then aggregates ... unless I can find a way to do incremental updates.

**Comment by velvia** [comment]
Since we are discussing streaming aggregations, I'd like to take the opportunity to point out my other OSS

-Evan
"Never doubt that a small group of thoughtful, committed citizens can change the world" - M. Mead

**Comment by velvia** [comment]
Oops, didn't finish my thought.  I'd like to point out my other OSS

-Evan
"Never doubt that a small group of thoughtful, committed citizens can change the world" - M. Mead

**Comment by velvia** [comment]
Check out my other OSS project, FiloDB (GitHub.com/TupleJump/FiloDB).  It allows you to dump event data into C and query them in an ad hoc way using Spark very efficiently, at Parquet performance levels.  Sorry for the other emails and thanks for the plug :)

-Evan
"Never doubt that a small group of thoughtful, committed citizens can change the world" - M. Mead

**Comment by Prasidhdh** [comment]
I am working on spark to get refreshed data. I have source data in form of Mysql tables. I am using jdbc connection to get those data in spark. Then, i am doing some filtering, mapping and finally join between two Rdds. And then apply action on that joined rdd. Now, my Program is sleeping (thread sleep) for some time period and during that time, i am deleting data from original source tables and inserting new data. So, after thread sleep again i am calling action on same joined rdd. so it is my second job on same rdds.

In a result i need merged data (before thread sleep joined result + after thread sleep joined result) so i can get deleted data and new inserted data.

In this case, spark uses sort shuffle mechanism and internally it creates datafiles and indexfiles for each shuffle map task. So, during second job when i am getting new data, i want to merge or append this new data with previously stored datafile. For that, i tried to append new data to datafile but,I do not have read or write access to previously created datafile and indexfile. so, how can i merge results of both of the joins?

Please help me to solve my problem.

Thanks in advance!

**Comment by velvia** [comment]
In general you don’t have access to SPark’s internal shuffle files.  You can however persist the result of the first join (in HDFS, in memory, etc.) and append the results of the second join, or however you wish to combine them.   Does this not work?

Also beware you are on Ooyala job server mailing list, but the project is at spark-jobserver/spark-jobserver.

**Comment by Prasidhdh** [comment]
@velvia Thank you for your response. But, i want to do it internally like merging datafiles.
